packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -607,6 +607,41 @@ class NavigatorObserver {
   void didStopUserGesture() { }
 }
 
+/// An abstract class that provides the API to host a navigator observer.
+///
+/// Only the [HeroController] should be hosted in this class. In general, this
+/// class should only be used in the [MaterialApp] or the [CupertinoApp]. If
+/// you want to subscribe any other navigator observers (including your own
+/// [HeroController]), uses the [Navigator.observers] instead.
+///
+/// See also:
+///
+///  * [Navigator.observers], which provides the standard way of subscribing
+///    navigator observers.
+class InheritedNavigatorObserver extends InheritedWidget {
```

Comment:
I cannot make it targeted because navigator.dart cannot depend on hero.dart or app.dart. otherwise it will be circular dependencies. The other choice is merge hero.dart and navigator.dart, but that will probably be a nasty breaking change.

Reply:
Why can navigator.dart and hero.dart not import each other? They are both in widgets layer.

Reply:
Is that acceptable? I thought that is not a good design pattern. Do we have example in our code base that does it?

Reply:
navigator.dart and routes.dart - within the same layer it should be fine.